File: ios/Sources/RoadSenseNSBootstrap/Pipeline/ReadingBuilder.swift
import Foundation

public struct LocationSample: Equatable, Sendable {
    public let timestamp: TimeInterval
    public let latitude: Double
    public let longitude: Double
    public let horizontalAccuracyMeters: Double
    public let speedKmh: Double
    public let headingDegrees: Double

    public init(
        timestamp: TimeInterval,
        latitude: Double,
        longitude: Double,
        horizontalAccuracyMeters: Double,
        speedKmh: Double,
        headingDegrees: Double
    ) {
        self.timestamp = timestamp
        self.latitude = latitude
        self.longitude = longitude
        self.horizontalAccuracyMeters = horizontalAccuracyMeters
        self.speedKmh = speedKmh
        self.headingDegrees = headingDegrees
    }
}

public struct ReadingWindow: Equatable, Sendable {
    public let latitude: Double
    public let longitude: Double
    public let roughnessRMS: Double
    public let speedKmh: Double
    public let headingDegrees: Double
    public let gpsAccuracyMeters: Double
    public let startedAt: TimeInterval
    public let durationSeconds: TimeInterval
    public let sampleCount: Int
    public let potholeSpikeCount: Int
    public let potholeMaxG: Double

    public init(
        latitude: Double,
        longitude: Double,
        roughnessRMS: Double,
        speedKmh: Double,
        headingDegrees: Double,
        gpsAccuracyMeters: Double,
        startedAt: TimeInterval,
        durationSeconds: TimeInterval,
        sampleCount: Int,
        potholeSpikeCount: Int,
        potholeMaxG: Double
    ) {
        self.latitude = latitude
        self.longitude = longitude
        self.roughnessRMS = roughnessRMS
        self.speedKmh = speedKmh
        self.headingDegrees = headingDegrees
        self.gpsAccuracyMeters = gpsAccuracyMeters
        self.startedAt = startedAt
        self.durationSeconds = durationSeconds
        self.sampleCount = sampleCount
        self.potholeSpikeCount = potholeSpikeCount
        self.potholeMaxG = potholeMaxG
    }
}

public struct ReadingBuilder: Sendable {
    public var targetDistanceMeters: Double = 40
    public var maxDurationSeconds: TimeInterval = 15
    public var maxHorizontalAccuracyMeters: Double = 20
    public var maxHeadingVarianceDegrees: Double = 60
    public var minimumSampleCount: Int = 30

    private var locationSamples: [LocationSample] = []
    private var motionSamples: [MotionSample] = []

    public init() {}

    public mutating func addMotionSample(_ sample: MotionSample) {
        motionSamples.append(sample)
    }

    public mutating func addLocationSample(_ sample: LocationSample) -> ReadingWindow? {
        guard sample.horizontalAccuracyMeters <= maxHorizontalAccuracyMeters else {
            reset()
            return nil
        }

        if locationSamples.isEmpty {
            locationSamples = [sample]
            return nil
        }

        locationSamples.append(sample)

        let duration = sample.timestamp - locationSamples[0].timestamp
        guard duration <= maxDurationSeconds else {
            reset(startingWith: sample)
            return nil
        }

        guard headingVarianceDegrees(for: locationSamples) <= maxHeadingVarianceDegrees else {
            reset(startingWith: sample)
            return nil
        }

        guard traveledDistanceMeters(for: locationSamples) >= targetDistanceMeters else {
            return nil
        }

        guard motionSamples.count >= minimumSampleCount else {
            reset(startingWith: sample)
            return nil
        }

        let reading = ReadingWindow(
            latitude: midpoint(for: locationSamples).latitude,
            longitude: midpoint(for: locationSamples).longitude,
            roughnessRMS: rms(of: motionSamples),
            speedKmh: averageSpeed(for: locationSamples),
            headingDegrees: weightedHeading(for: locationSamples),
            gpsAccuracyMeters: locationSamples.map(\.horizontalAccuracyMeters).max() ?? sample.horizontalAccuracyMeters,
            startedAt: locationSamples[0].timestamp,
            durationSeconds: duration,
            sampleCount: motionSamples.count,
            potholeSpikeCount: 0,
            potholeMaxG: 0
        )

        reset(startingWith: sample)
        return reading
    }

    private mutating func reset(startingWith sample: LocationSample? = nil) {
        locationSamples = sample.map { [$0] } ?? []
        motionSamples = []
    }

    private func rms(of samples: [MotionSample]) -> Double {
        let meanSquare = samples.reduce(0.0) { partialResult, sample in
            let value = sample.verticalAcceleration
            return partialResult + (value * value)
        } / Double(samples.count)

        return sqrt(meanSquare)
    }

    private func averageSpeed(for samples: [LocationSample]) -> Double {
        samples.map(\.speedKmh).reduce(0, +) / Double(samples.count)
    }

    private func midpoint(for samples: [LocationSample]) -> (latitude: Double, longitude: Double) {
        let totalDistance = traveledDistanceMeters(for: samples)
        let target = totalDistance / 2

        guard totalDistance > 0, samples.count > 1 else {
            let sample = samples[samples.count / 2]
            return (sample.latitude, sample.longitude)
        }

        var traversed = 0.0

        for index in 1..<samples.count {
            let previous = samples[index - 1]
            let current = samples[index]
            let segmentDistance = distanceMeters(from: previous, to: current)

            if traversed + segmentDistance >= target, segmentDistance > 0 {
                let progress = (target - traversed) / segmentDistance
                return (
                    latitude: previous.latitude + ((current.latitude - previous.latitude) * progress),
                    longitude: previous.longitude + ((current.longitude - previous.longitude) * progress)
                )
            }

            traversed += segmentDistance
        }

        let last = samples.last!
        return (last.latitude, last.longitude)
    }

    private func traveledDistanceMeters(for samples: [LocationSample]) -> Double {
        guard samples.count > 1 else {
            return 0
        }

        return zip(samples, samples.dropFirst()).reduce(0.0) { partialResult, pair in
            partialResult + distanceMeters(from: pair.0, to: pair.1)
        }
    }

    private func distanceMeters(from lhs: LocationSample, to rhs: LocationSample) -> Double {
        let earthRadiusMeters = 6_371_000.0
        let lat1 = lhs.latitude * .pi / 180
        let lat2 = rhs.latitude * .pi / 180
        let deltaLat = (rhs.latitude - lhs.latitude) * .pi / 180
        let deltaLon = (rhs.longitude - lhs.longitude) * .pi / 180

        let a = sin(deltaLat / 2) * sin(deltaLat / 2) +
            cos(lat1) * cos(lat2) * sin(deltaLon / 2) * sin(deltaLon / 2)
        let c = 2 * atan2(sqrt(a), sqrt(1 - a))

        return earthRadiusMeters * c
    }

    private func weightedHeading(for samples: [LocationSample]) -> Double {
        let weights = samples.map { max($0.speedKmh, 0.1) }
        let x = zip(samples, weights).reduce(0.0) { partialResult, pair in
            partialResult + cos(pair.0.headingDegrees * .pi / 180) * pair.1
        }
        let y = zip(samples, weights).reduce(0.0) { partialResult, pair in
            partialResult + sin(pair.0.headingDegrees * .pi / 180) * pair.1
        }

        let angle = atan2(y, x) * 180 / .pi
        return angle >= 0 ? angle : angle + 360
    }

    private func headingVarianceDegrees(for samples: [LocationSample]) -> Double {
        let mean = weightedHeading(for: samples)

        return samples.reduce(0.0) { partialResult, sample in
            max(partialResult, angularDistanceDegrees(from: sample.headingDegrees, to: mean))
        }
    }

    private func angularDistanceDegrees(from lhs: Double, to rhs: Double) -> Double {
        let raw = abs(lhs - rhs).truncatingRemainder(dividingBy: 360)
        return min(raw, 360 - raw)
    }
}

File: android/core-sensor/src/main/kotlin/ca/roadsense/ns/sensor/ReadingBuilder.kt
package ca.roadsense.ns.sensor

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

data class ReadingWindow(
    val latitude: Double,
    val longitude: Double,
    val roughnessRMS: Double,
    val speedKmh: Double,
    val headingDegrees: Double,
    val gpsAccuracyMeters: Double,
    val startedAt: Double,
    val durationSeconds: Double,
    val sampleCount: Int,
    val potholeSpikeCount: Int,
    val potholeMaxG: Double,
)

class ReadingBuilder(
    var targetDistanceMeters: Double = 40.0,
    var maxDurationSeconds: Double = 15.0,
    var maxHorizontalAccuracyMeters: Double = 20.0,
    var maxHeadingVarianceDegrees: Double = 60.0,
    var minimumSampleCount: Int = 30,
    var roughnessScorer: RoughnessScorer = RoughnessScorer(),
) {
    data class Snapshot(
        val targetDistanceMeters: Double,
        val maxDurationSeconds: Double,
        val maxHorizontalAccuracyMeters: Double,
        val maxHeadingVarianceDegrees: Double,
        val minimumSampleCount: Int,
        val locationSamples: List<LocationSample>,
        val motionSamples: List<MotionSample>,
    )

    private val locationSamples: MutableList<LocationSample> = mutableListOf()
    private val motionSamples: MutableList<MotionSample> = mutableListOf()

    constructor(snapshot: Snapshot) : this(
        targetDistanceMeters = snapshot.targetDistanceMeters,
        maxDurationSeconds = snapshot.maxDurationSeconds,
        maxHorizontalAccuracyMeters = snapshot.maxHorizontalAccuracyMeters,
        maxHeadingVarianceDegrees = snapshot.maxHeadingVarianceDegrees,
        minimumSampleCount = snapshot.minimumSampleCount,
    ) {
        locationSamples.addAll(snapshot.locationSamples)
        motionSamples.addAll(snapshot.motionSamples)
    }

    fun addMotionSample(sample: MotionSample) {
        motionSamples.add(sample)
    }

    fun addLocationSample(sample: LocationSample): ReadingWindow? {
        if (sample.horizontalAccuracyMeters > maxHorizontalAccuracyMeters) {
            reset()
            return null
        }

        if (locationSamples.isEmpty()) {
            locationSamples.add(sample)
            return null
        }

        locationSamples.add(sample)

        val duration = sample.timestamp - locationSamples[0].timestamp
        if (duration > maxDurationSeconds) {
            reset(startingWith = sample)
            return null
        }

        if (headingVarianceDegrees(locationSamples) > maxHeadingVarianceDegrees) {
            reset(startingWith = sample)
            return null
        }

        if (traveledDistanceMeters(locationSamples) < targetDistanceMeters) {
            return null
        }

        if (motionSamples.size < minimumSampleCount) {
            reset(startingWith = sample)
            return null
        }

        val midpoint = midpoint(locationSamples)
        val window = ReadingWindow(
            latitude = midpoint.first,
            longitude = midpoint.second,
            roughnessRMS = roughnessScorer.score(motionSamples),
            speedKmh = averageSpeed(locationSamples),
            headingDegrees = weightedHeading(locationSamples),
            gpsAccuracyMeters = locationSamples.maxOf { it.horizontalAccuracyMeters },
            startedAt = locationSamples[0].timestamp,
            durationSeconds = duration,
            sampleCount = motionSamples.size,
            potholeSpikeCount = 0,
            potholeMaxG = 0.0,
        )

        reset(startingWith = sample)
        return window
    }

    fun snapshot(): Snapshot = Snapshot(
        targetDistanceMeters = targetDistanceMeters,
        maxDurationSeconds = maxDurationSeconds,
        maxHorizontalAccuracyMeters = maxHorizontalAccuracyMeters,
        maxHeadingVarianceDegrees = maxHeadingVarianceDegrees,
        minimumSampleCount = minimumSampleCount,
        locationSamples = locationSamples.toList(),
        motionSamples = motionSamples.toList(),
    )

    private fun reset(startingWith: LocationSample? = null) {
        locationSamples.clear()
        if (startingWith != null) locationSamples.add(startingWith)
        motionSamples.clear()
    }

    private fun averageSpeed(samples: List<LocationSample>): Double =
        samples.sumOf { it.speedKmh } / samples.size

    private fun midpoint(samples: List<LocationSample>): Pair<Double, Double> {
        val totalDistance = traveledDistanceMeters(samples)
        val target = totalDistance / 2

        if (totalDistance <= 0 || samples.size <= 1) {
            val sample = samples[samples.size / 2]
            return sample.latitude to sample.longitude
        }

        var traversed = 0.0
        for (index in 1 until samples.size) {
            val previous = samples[index - 1]
            val current = samples[index]
            val segmentDistance = distanceMeters(previous, current)

            if (traversed + segmentDistance >= target && segmentDistance > 0) {
                val progress = (target - traversed) / segmentDistance
                return Pair(
                    previous.latitude + ((current.latitude - previous.latitude) * progress),
                    previous.longitude + ((current.longitude - previous.longitude) * progress),
                )
            }
            traversed += segmentDistance
        }

        val last = samples.last()
        return last.latitude to last.longitude
    }

    private fun traveledDistanceMeters(samples: List<LocationSample>): Double {
        if (samples.size <= 1) return 0.0
        var total = 0.0
        for (i in 1 until samples.size) {
            total += distanceMeters(samples[i - 1], samples[i])
        }
        return total
    }

    private fun distanceMeters(lhs: LocationSample, rhs: LocationSample): Double {
        val earthRadiusMeters = 6_371_000.0
        val lat1 = lhs.latitude * PI / 180
        val lat2 = rhs.latitude * PI / 180
        val deltaLat = (rhs.latitude - lhs.latitude) * PI / 180
        val deltaLon = (rhs.longitude - lhs.longitude) * PI / 180

        val a = sin(deltaLat / 2) * sin(deltaLat / 2) +
            cos(lat1) * cos(lat2) * sin(deltaLon / 2) * sin(deltaLon / 2)
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return earthRadiusMeters * c
    }

    private fun weightedHeading(samples: List<LocationSample>): Double {
        var x = 0.0
        var y = 0.0
        for (sample in samples) {
            val weight = if (sample.speedKmh > 0.1) sample.speedKmh else 0.1
            x += cos(sample.headingDegrees * PI / 180) * weight
            y += sin(sample.headingDegrees * PI / 180) * weight
        }
        val angle = atan2(y, x) * 180 / PI
        return if (angle >= 0) angle else angle + 360
    }

    private fun headingVarianceDegrees(samples: List<LocationSample>): Double {
        val mean = weightedHeading(samples)
        var maxDelta = 0.0
        for (sample in samples) {
            val delta = angularDistanceDegrees(sample.headingDegrees, mean)
            if (delta > maxDelta) maxDelta = delta
        }
        return maxDelta
    }

    private fun angularDistanceDegrees(lhs: Double, rhs: Double): Double {
        val raw = abs(lhs - rhs).mod(360.0)
        return min(raw, 360 - raw)
    }
}
